### PR TITLE
Added validation for duplicated network port for logging receivers

### PR DESCRIPTION
**Pull request by LujieDuan**
[b/220128869](http://b/220128869) | P3 | Add validation failure at the Ops Agent level when there are two receivers listening on the same port
The validation added here can check for both cases: 1. two receivers using the same port and 2. one receiver used by two pipelines. 
Originally considered to use custom validator but that would require traverse the fields in the yaml tree between `service` and `receivers` sections to check receivers' port and whether the receiver is used by any pipeline. 
This current proposed solution added a new interface LoggingNetworkReceiver by extending the LoggingReceiver interface and provide the method to access the port taken by this receiver. The the validation is actually happening in the `Validate()` on the `Logging` instance.  
To validate, loop over all services/pipelines and keep track of ports used, and return the error if there is port conflict. 
Hostname is not considered in this validation; i.e., if two syslog receivers listening on `127.0.0.1:5540` and `0.0.0.0:5540` we still consider there is conflict.

### CODE REVIEW

File: confgenerator/testdata/invalid/linux/logging-receiver_duplicated_port/input.yaml [hunk 1]
```diff
@@ -0,0 +1,18 @@
+logging:
```
Comment (qingling128) [inline confgenerator/testdata/invalid/linux/logging-receiver_duplicated_port/input.yaml:1]: Can we also add a unit test case like this for completeness?

```
logging:
  receivers:
    s1:
      type: syslog
      transport_protocol: tcp
      listen_host: 0.0.0.0
      listen_port: 5140      
    s2:
      type: syslog
      transport_protocol: tcp
      listen_host: 0.0.0.0
      listen_port: 5140  
  service:
    pipelines:
      p1:
        receivers: [s1, s2]
```

File: confgenerator/config.go [hunk 2]
```diff
@@ -865,6 +882,32 @@ func validateComponentTypeCounts(components interface{}, refs []string, subagent
 	return r, nil
 }
 
+// Validate that no two receivers are using the same port; adding new port usage to the input map `taken`
+func validateReceiverPorts(taken map[uint16]string, components interface{}, refs []string) (map[uint16]string, error) {
```
Comment (qingling128) [inline confgenerator/config.go:886]: [Optional] I'm trying to think how to better distinguish the IDs from `refs` and the IDs from `taken[port]` to help with readability by naming them in a more descriptive way.

Some brainstorming ideas are: `pipelineRIDs` and `portRID`. e.g.

```
for _, pipelineRID := range pipelineRIDs {
...
   if portRID, ok := taken[port]; ok { 
        if portRID == pipelineRID
```
But they might be long per Golang naming standards. Will defer to you whether to make any adjustment.

File: confgenerator/config.go [hunk 3]
```diff
@@ -776,6 +787,12 @@ var (
 		"iis":                     1,
 		"mssql":                   1,
 	}
+
+	receiverPortLimits = map[string]bool{
```
[the remark this answers is not in the document]
Comment (LujieDuan) [reply confgenerator/config.go]: Good point! Replaced this with a slice and updated the validation function to loop over the slice to find if the type is in this slice.

File: confgenerator/config.go (hunk 2 above)
Comment (LujieDuan) [reply confgenerator/config.go]: Yes agree those are not good names here to distinguish those ids. Updated the naming as suggested.

File: confgenerator/testdata/invalid/linux/logging-receiver_duplicated_port/input.yaml (hunk 1 above)
Comment (LujieDuan) [reply confgenerator/testdata/invalid/linux/logging-receiver_duplicated_port/input.yaml:1]: Test case added as `logging-receiver_duplicated_port_same_service`

**Review by qingling128 — APPROVED** [review]
LGTM

**Pull request merged**